index.js

```diff
@@ -38,110 +28,54 @@ const defaultOptions = {
 
 module.exports = function prettyFactory (options) {
   const opts = Object.assign({}, defaultOptions, options)
-  const EOL = opts.crlf ? '\r\n' : '\n'
-  const IDENT = '    '
   const messageKey = opts.messageKey
   const timestampKey = opts.timestampKey
-  const errorLikeObjectKeys = opts.errorLikeObjectKeys
-  const errorProps = opts.errorProps.split(',')
-  const ignoreKeys = opts.ignore ? new Set(opts.ignore.split(',')) : undefined
 
   const colorizer = colors(opts.colorize)
-  const search = opts.search
-
-  return pretty
 
-  function pretty (inputData) {
-    let log
-    if (!isObject(inputData)) {
-      const parsed = jsonParser(inputData)
-      log = parsed.value
-      if (parsed.err) {
-        // pass through
-        return inputData + EOL
-      }
-    } else {
-      log = inputData
-    }
-
-    // Short-circuit for spec allowed primitive values.
-    if ([null, true, false].includes(log) || Number.isFinite(log)) {
-      return `${log}\n`
-    }
-
-    if (search && !jmespath.search(log, search)) {
-      return
-    }
-
-    if (ignoreKeys) {
-      log = Object.keys(log)
-        .filter(key => !ignoreKeys.has(key))
-        .reduce((res, key) => {
-          res[key] = log[key]
-          return res
-        }, {})
-    }
+  const context = {
+    opts,
+    EOL: opts.crlf ? '\r\n' : '\n',
+    IDENT: '    '
+  }
 
-    const prettifiedLevel = prettifyLevel({ log, colorizer })
-    const prettifiedMessage = prettifyMessage({ log, messageKey, colorizer })
-    const prettifiedMetadata = prettifyMetadata({ log })
-    const prettifiedTime = prettifyTime({ log, translateFormat: opts.translateTime, timestampKey })
+  if (opts.logParsers) {
+    logParsers.push(...opts.logParsers)
+  }
 
-    let line = ''
-    if (opts.levelFirst && prettifiedLevel) {
-      line = `${prettifiedLevel}`
-    }
+  if (opts.lineBuilders) {
+    lineBuilders.push(...opts.lineBuilders)
+  }
 
-    if (prettifiedTime && line === '') {
-      line = `${prettifiedTime}`
-    } else if (prettifiedTime) {
-      line = `${line} ${prettifiedTime}`
-    }
+  return pretty
 
-    if (!opts.levelFirst && prettifiedLevel) {
-      if (line.length > 0) {
-        line = `${line} ${prettifiedLevel}`
-      } else {
-        line = prettifiedLevel
+  function pretty (inputData) {
+    let nextInput = inputData
+
+    for (const logParser of logParsers) {
```

Comment:
A major concern I have with the rewrite of this file is the introduction of a lot of iteration. You iterate over "log parsers" which in turn iterates over "line builders". I bet if we had some benchmarks, and that's a failing on our part, we would see a significant difference in processing time with this method over the very linear method currently on `master`. Prettification is expensive.

Reply:
It's definitely a valid concern. I found an interesting benchmark (https://jsperf.com/loops/285) that compares the performance of different kinds of loops and discovered, to my surprise, that the `for of` loop is pretty slow. I've changed the loops to classic `for` loops, which are either the fastest loops or among the fastest loops, though I'm not sure if the benchmark is relevant for node on the desktop.

Having a benchmark for pino-pretty would be useful. But even without it, it's important to remember that we're talking about the prettifier and not the logger itself. Performance is extremely important but I do think that mitigates it a bit.

Finally, one of the advantages of this approach is that the prettifier can be optimized very easily by simply not using the parsers and builders that are irrelevant for an app. For example, if you don't want to show the timestamp, don't use the time log processor, thus avoiding the parsing, prettification, and building costs associated with it.

Reply:
> But even without it, it's important to remember that we're talking about the prettifier and not the logger itself. 

Agreed. In particular when running as a piped process. But it still a concern nonetheless.